lib/cutting_edge/workers/dependency.rb

```diff
@@ -28,14 +29,66 @@ def perform(identifier, lang, locations, dependency_types, to_addr, auth_token =
       add_to_store(identifier, dependencies) unless @nothing_changed
     ensure
       unless @nothing_changed
+      #  alt_diff(dependencies[:locations], old_dependencies[:locations])
         badge_worker(identifier)
-        mail_worker(identifier, to_addr) if to_addr
+        mail_worker(identifier, to_addr, alt_diff(old_dependencies[:locations], dependencies[:locations])) if to_addr && !old_dependencies.empty?
       end
       GC.start
     end
   end
 
   private
+  
+  def alt_diff(old, new)
+    old_merged = old.values.first.merge(*old.values[1..-1]) {|key, current_val, new_val| current_val + new_val}
+    new_merged = new.values.first.merge(*new.values[1..-1]) {|key, current_val, new_val| current_val + new_val}
+    diff = Hashdiff.diff(old_merged, new_merged)
+    additions = diff.select {|x| x.first == '+'}
+    deletions = diff.select {|x| x.first == '-'}
+    result = {}
+    
+    additions.each do |addition|
+      status = addition[1].split('[').first.to_sym # Hashdiff generates strings of the form "ok[1]" to indicate this is the first change to the Hash under the :ok key
+      name = addition.last[:name]
+      old_status = deletions.find {|x| x.last[:name] == name}
+      old_status = old_status[1].split('[').first.to_sym if old_status
+      result[name] = status_improved?(old_status, status) ? :good_change : :bad_change
+    end
+    result
+  end
+  
+  def diff_dependencies(old, new)
+    old_with_status = status_for_all_dependencies(old)
+    new_with_status = status_for_all_dependencies(new)
+    
+    new_with_status.delete_if {|k,v| old_with_status[k] == v}
+    
+    new_with_status.each do |name, status|
+      new_with_status[name] = status_improved?(old_with_status[name], status) ? :good_change : :bad_change
+    end
+    new_with_status
+  end
+  
+  def status_improved?(old_status, new_status)
```

Comment:
I'm wondering if this method shouldn't return `good_change` or `bad_change` instead (this would mean changing the method signature).